### ISSUE TITLE
Installation issue: pflogger: MPI is picked up and enabled when variant is set to false

### ISSUE THREAD

**Issue by climbfuji** [LLNL/spack#42137]
### Steps to reproduce the issue

pflogger has a variant `mpi` which is set to false (`~mpi`) by default. But if for some reason a user has SOME version of MPI in the default search path (as it is the case on an AWS ParallelCluster image, `which mpiexec` --> `/opt/amazon/openmpi/bin/mpiexec`), then the pflogger build picks it up:
```
# MPI support will be built if FindPackage() finds MPI in the build
# environment.
...
find_package (MPI QUIET)
```
This leads to build errors if the stack being built uses a different compiler than the one used to compile that default MPI.

### Error message

```
==> Installing pflogger-1.10.0-tfq4l6mjzpu67mehlwenb5li3uqlapsq [45/54]
==> No binary for pflogger-1.10.0-tfq4l6mjzpu67mehlwenb5li3uqlapsq found: installing from source
==> Warning: Intel's compilers may or may not optimize to the same degree for non-Intel microprocessors for optimizations that are not unique to Intel microprocessors
==> Fetching file:///mnt/experiments-efs/spack-stack/source-cache/_source-cache/archive/8e/8e25564699c0adcbe9a23fded6637668ce659480b39420be5a4c8181cd44ad53.tar.gz
==> No patches needed for pflogger
==> pflogger: Executing phase: 'cmake'
==> [2024-01-18-03:05:54.430731] '/mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/envs/unified-env/install/intel/2022.1.0/cmake-3.23.1-k3psmwy/bin/cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/envs/unified-env/install/intel/2022.1.0/pflogger-1.10.0-tfq4l6m' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON' '-DCMAKE_INSTALL_RPATH:STRING=/mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/envs/unified-env/install/intel/2022.1.0/pflogger-1.10.0-tfq4l6m/lib;/mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/envs/unified-env/install/intel/2022.1.0/pflogger-1.10.0-tfq4l6m/lib64' '-DCMAKE_PREFIX_PATH:STRING=/mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/envs/unified-env/install/intel/2022.1.0/yafyaml-1.1.0-oyxcgd3;/mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/envs/unified-env/install/intel/2022.1.0/gftl-shared-1.6.1-cy67ujq;/mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/envs/unified-env/install/intel/2022.1.0/gftl-1.10.0-nlo6xs6;/mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/envs/unified-env/install/intel/2022.1.0/cmake-3.23.1-k3psmwy' '/mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/cache/build_stage/spack-stage-pflogger-1.10.0-tfq4l6mjzpu67mehlwenb5li3uqlapsq/spack-src'
-- The Fortran compiler identification is Intel 2021.6.0.20220226
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/spack/lib/spack/env/intel/ifort - skipped
-- Performing Test SUPPORT_FOR_ASSUMED_TYPE
-- Performing Test SUPPORT_FOR_ASSUMED_TYPE: SUCCESS
-- Performing Test SUPPORT_FOR_C_LOC_ASSUMED_SIZE
-- Performing Test SUPPORT_FOR_C_LOC_ASSUMED_SIZE: FAILURE
-- Performing Test SUPPORT_FOR_MPI_ALLOC_MEM_CPTR
-- Performing Test SUPPORT_FOR_MPI_ALLOC_MEM_CPTR: FAILURE
-- Performing Test _ISO_REAL32: SUCCESS (value=4)
-- Performing Test _ISO_REAL64: SUCCESS (value=8)
-- Performing Test _ISO_REAL128: SUCCESS (value=16)
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
-- Configuring done
-- Generating done
-- Build files have been written to: /mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/cache/build_stage/spack-stage-pflogger-1.10.0-tfq4l6mjzpu67mehlwenb5li3uqlapsq/spack-build-tfq4l6m
==> pflogger: Executing phase: 'build'

...

  >> 93     /mnt/experiments-zfs/dom.heinzeller/spack-stack-geos-dev/spack-stack/cache/build_stage/spack-stage-pflogger-1.10.0-tfq4l6mjzpu67mehlwenb5li3uqlapsq/spack-src/src/MpiCommConfig.F90(3): error #7013: This module file was not generated by any release
            of this compiler.   [MPI]
     94        use mpi
     95     -------^
```

There are two issues I think:
1. `spack doesn't fully sanitize the environment and `mpiexec`/`mpif90` remain in the path *OR* it is expected that no MPI is in the default search path but it is. (@scheibelp or @tldahlgren can you comment on this please?)
2. `pflogger` should not enable `MPI` based on hidden/quiet `cmake` `find_package` commands but should have an explicit switch for it.

### Information on your system

```
* **Spack:** 0.21.1.dev0 (2d4531c8e6670f3f3e17297f502d5de81cc14e23)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-zen2
* **Concretizer:** clingo
```

### Additional information

[spack-logs.tar.gz](https://github.com/spack/spack/files/13971709/spack-logs.tar.gz)

@mathomp4  @tclune

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by tclune** [comment]
Even if I add an explicit switch to activate the MPI components of pFlogger, won't the `find_package()` command still find the default MPI?    AFAIK, all extant users of pFlogger will be using it _with_ MPI.      

Basically I'm asking a question about my ignorance?  Are there extra arguments to `find_package` that should be added to direct it to the non-default MPI?     I've just always arranged my environment to have the desired MPI in my path ...

**Comment by climbfuji** [comment]
> Even if I add an explicit switch to activate the MPI components of pFlogger, won't the `find_package()` command still find the default MPI? AFAIK, all extant users of pFlogger will be using it _with_ MPI.
> 
> Basically I'm asking a question about my ignorance? Are there extra arguments to `find_package` that should be added to direct it to the non-default MPI? I've just always arranged my environment to have the desired MPI in my path ...

```
if (ENABLE_MPI)
    find_package(MPI REQUIRED)
endif()
```
makes sure you don't find MPI if `-DENABLE_MPI` is `OFF`. And spack makes sure (see `package.py`) that the correct MPI is loaded when using variant `+mpi`.

**Comment by tclune** [comment]
I'm happy to ad a `DISABLE_MPI` option:
```
if (NOT DISABLE_MPI)
    find_package(MPI REQUIRED)
endif()
```

I had conflated the badness of a "positive" default with the issue you are really having here.

**Comment by climbfuji** [comment]
I think @mathomp4 summarized it well, use `ENABLE_MPI` and make it default to on (same as the variant in spack).

**Comment by scheibelp** [comment]
Regarding

> spack doesn't fully sanitize the environment and `mpiexec`/`mpif90` remain in the path OR it is expected that no MPI is in the default search path but it is.

To confirm: we don't remove entries from the user's PATH: if the user adds a directory, and a package looks for an executable, we will use whatever is in PATH (we tend to get what we want though because Spack inserts Spack-installed packages first into the PATH - that doesn't help here though).

This `ENABLE_MPI`  (i.e. disabling `find_package` when `ENABLE_MPI=off`) route would be great for Spack.

Longer-term, I think it would be possible to improve our isolation: we only depend on PATH for some common things in `/usr/bin` (things like `git`, `file`, and `make`) and we could create a directory that symlinks to this limited set of utilities (or I suppose the exact strategy isn't important: more that we could avoid pulling in unexpected executables).

EDIT: It was later pointed out to me that we already have https://github.com/spack/spack/pull/30015, so I plan to look at that next week.

**Comment by climbfuji** [comment]
Thanks for the info @scheibelp! Yes, we already made the changes in pflogger to use `ENABLE_MPI` - will be available with the next tag I think.

**Comment by mathomp4** [comment]
> Thanks for the info @scheibelp! Yes, we already made the changes in pflogger to use `ENABLE_MPI` - will be available with the next tag I think.

Indeed, pFlogger v1.12 was released today and #42288 should add it to spack.

**Comment by climbfuji** [comment]
Closed via https://github.com/spack/spack/pull/42288
